File: painterly/flowmap.py
"""灰度梯度 → 法线图 / 流场（笔触方向场）。

关键不变量：笔触方向 = 边缘切线方向（isophote），是 π 周期的"朝向"而非向量，
平滑和重采样必须在结构张量（等价于 2θ 复数域）上做，直接对角度求平均会在
0/π 交界处出错。
"""
import numpy as np
import cv2


def gradients(gray: np.ndarray):
    gx = cv2.Sobel(gray, cv2.CV_32F, 1, 0, ksize=3)
    gy = cv2.Sobel(gray, cv2.CV_32F, 0, 1, ksize=3)
    return gx, gy


def normal_map(gx: np.ndarray, gy: np.ndarray, strength: float = 4.0) -> np.ndarray:
    """把模糊灰度当高度场生成切线空间法线图（所以整体偏蓝紫）。"""
    nx, ny, nz = -gx * strength, gy * strength, np.ones_like(gx)
    norm = np.sqrt(nx * nx + ny * ny + nz * nz)
    n = np.stack([nx / norm, ny / norm, nz / norm], axis=-1)
    return ((n * 0.5 + 0.5) * 255).astype(np.uint8)


def flow_field(gx: np.ndarray, gy: np.ndarray, smooth_sigma: float = 6.0):
    """结构张量平滑后的方向场。

    返回 (theta, coherence)：
      theta      — 每个像素的笔触方向（边缘切线，弧度，π 周期）
      coherence  — 方向的可信度 0..1（各向异性程度），平坦区接近 0
    """
    jxx = cv2.GaussianBlur(gx * gx, (0, 0), smooth_sigma)
    jxy = cv2.GaussianBlur(gx * gy, (0, 0), smooth_sigma)
    jyy = cv2.GaussianBlur(gy * gy, (0, 0), smooth_sigma)

    # 结构张量主特征向量方向 = 平滑后的梯度朝向；笔触沿其垂直方向
    phi = 0.5 * np.arctan2(2 * jxy, jxx - jyy)
    theta = phi + np.pi / 2

    tmp = np.sqrt((jxx - jyy) ** 2 + 4 * jxy * jxy)
    lam1, lam2 = (jxx + jyy + tmp) / 2, (jxx + jyy - tmp) / 2
    coherence = np.where(lam1 + lam2 > 1e-8, (lam1 - lam2) / (lam1 + lam2 + 1e-8), 0.0)
    return theta.astype(np.float32), coherence.astype(np.float32)


def flow_map_vis(theta: np.ndarray, coherence: np.ndarray) -> np.ndarray:
    """流场可视化。方向是 π 周期的，用 2θ 编码进 R/G；
    压低摆幅、抬高基底，得到橄榄绿基调的 flow map 贴图观感。"""
    r = np.cos(2 * theta) * 0.22 * (0.3 + 0.7 * coherence) + 0.52
    g = np.sin(2 * theta) * 0.22 * (0.3 + 0.7 * coherence) + 0.50
    b = coherence * 0.18 + 0.06
    return (np.clip(np.stack([r, g, b], axis=-1), 0, 1) * 255).astype(np.uint8)

File: crates/painterly-core/src/flowmap.rs
//! グレースケール勾配 → 法線マップ / フローマップ（ストロークの方向場）。
//!
//! 不変条件: ストローク方向 = エッジの接線方向（等輝度線）で、π 周期の「向き」
//! であってベクトルではない。平滑化・リサンプリングは構造テンソル
//! （= 2θ の複素数域）上で行うこと。角度を直接平均すると 0/π の境目で壊れる。
//! Python 版 painterly/flowmap.py の移植。

use crate::buf::{gaussian_blur, Gray};

/// 高さ場と見なした勾配から接線空間法線マップを生成（全体が青紫になる）
pub fn normal_map(gx: &Gray, gy: &Gray, strength: f32) -> image::RgbImage {
    let (w, h) = (gx.w, gx.h);
    let mut out = image::RgbImage::new(w as u32, h as u32);
    for (i, p) in out.pixels_mut().enumerate() {
        let nx = -gx.data[i] * strength;
        let ny = gy.data[i] * strength;
        let nz = 1.0f32;
        let norm = (nx * nx + ny * ny + nz * nz).sqrt();
        p.0 = [
            ((nx / norm * 0.5 + 0.5) * 255.0) as u8,
            ((ny / norm * 0.5 + 0.5) * 255.0) as u8,
            ((nz / norm * 0.5 + 0.5) * 255.0) as u8,
        ];
    }
    out
}

/// 構造テンソルを平滑化した方向場。
/// 戻り値 (theta, coherence):
///   theta     — 各画素のストローク方向（エッジ接線、ラジアン、π 周期）
///   coherence — 方向の信頼度 0..1（異方性の強さ）。平坦部はほぼ 0
pub fn flow_field(gx: &Gray, gy: &Gray, smooth_sigma: f32) -> (Gray, Gray) {
    let (w, h) = (gx.w, gx.h);
    let mut jxx = Gray::new(w, h);
    let mut jxy = Gray::new(w, h);
    let mut jyy = Gray::new(w, h);
    for i in 0..w * h {
        jxx.data[i] = gx.data[i] * gx.data[i];
        jxy.data[i] = gx.data[i] * gy.data[i];
        jyy.data[i] = gy.data[i] * gy.data[i];
    }
    let jxx = gaussian_blur(&jxx, smooth_sigma);
    let jxy = gaussian_blur(&jxy, smooth_sigma);
    let jyy = gaussian_blur(&jyy, smooth_sigma);

    let mut theta = Gray::new(w, h);
    let mut coherence = Gray::new(w, h);
    for i in 0..w * h {
        // 構造テンソルの主固有ベクトル方向 = 平滑化された勾配の向き。
        // ストロークはその垂直方向に走らせる
        let phi = 0.5 * (2.0 * jxy.data[i]).atan2(jxx.data[i] - jyy.data[i]);
        theta.data[i] = phi + std::f32::consts::FRAC_PI_2;
        let tmp = ((jxx.data[i] - jyy.data[i]).powi(2) + 4.0 * jxy.data[i] * jxy.data[i]).sqrt();
        let lam1 = (jxx.data[i] + jyy.data[i] + tmp) / 2.0;
        let lam2 = (jxx.data[i] + jyy.data[i] - tmp) / 2.0;
        coherence.data[i] = if lam1 + lam2 > 1e-8 {
            (lam1 - lam2) / (lam1 + lam2 + 1e-8)
        } else {
            0.0
        };
    }
    (theta, coherence)
}

/// フローマップの可視化。方向は π 周期なので 2θ を R/G にエンコードし、
/// 振幅を抑えて下駄を履かせるとオリーブ色基調の flow map らしい見た目になる
pub fn flow_map_vis(theta: &Gray, coherence: &Gray) -> image::RgbImage {
    let (w, h) = (theta.w, theta.h);
    let mut out = image::RgbImage::new(w as u32, h as u32);
    for (i, p) in out.pixels_mut().enumerate() {
        let coh = coherence.data[i];
        let r = (2.0 * theta.data[i]).cos() * 0.22 * (0.3 + 0.7 * coh) + 0.52;
        let g = (2.0 * theta.data[i]).sin() * 0.22 * (0.3 + 0.7 * coh) + 0.50;
        let b = coh * 0.18 + 0.06;
        p.0 = [
            (r.clamp(0.0, 1.0) * 255.0) as u8,
            (g.clamp(0.0, 1.0) * 255.0) as u8,
            (b.clamp(0.0, 1.0) * 255.0) as u8,
        ];
    }
    out
}
